Traditionally statisticians and econometricians have been using statistical approaches to solve problems such as regression and classification. There are lot of similarities and differences between the statistical approach and machine learning (optimization approach [link](https://github.com/chidamnat/practicalDS/blob/master/mastery/Least_squares_regression.ipynb)).
We have seen the bit of ML approach earlier in the link above and there are libraries such as sklearn and others. More to come on this later

In this post let us take a peek into this using a popular library in python space is statsmodels which was inspired by statistical programming language R.

Before getting into this there are some pre-requisites.

1) [ML based approach for regression](https://github.com/chidamnat/practicalDS/blob/master/mastery/Least_squares_regression.ipynb)

2) [Hypothesis Testing](https://github.com/chidamnat/practicalDS/blob/master/mastery/Test_of_hypothesis.ipynb)

3) [Central Limit Theorem](https://github.com/chidamnat/practicalDS/blob/master/mastery/central_limit_theorem.ipynb) (we refer to standard_error here)

Let us use the same Advertising data and see if we could predict sales based on the multiple predictors(variables) this time.



In [0]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [0]:
df = pd.read_csv('Advertising.csv', index_col=0)
df.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

### Fit regression model

In [0]:
# Let us pick TV as a predictor to start with predicting the sales
result1 = smf.ols('sales ~ TV ', data=df).fit()
result1.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Sun, 15 Dec 2019   Prob (F-statistic):           1.47e-42
Time:                        15:32:36   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0326      0.458     15.360      0.000       6.130       7.935
TV             0.0475      0.003     17.668      0.000       0.042       0.053
==============================================================================
Omnibus:                        0.531   Durbin-Watson:                   1.935
Prob(Omnibus):                  0.767   Jarque-Bera (JB):                0.669
Skew:                          -0.089   Prob(JB):                        0.716
Kurtosis:                       2.779   Cond. No.                         338.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### This simple linear regression model explains 61.2% (from R-squared value) of the total variance. Which also means around 38.8% is still unexplained.

We can also see that the p-value associated with this predictor (TV) on sale is close to 0. So the model is pretty confident on the estimate of this coeffient.

To understand this behind the scenes, there is a [hypothesis testing](https://github.com/chidamnat/practicalDS/blob/master/mastery/Test_of_hypothesis.ipynb) involved. Null Hypothesis here is that there is no significant linear relationship between the variable TV and sales, which is coeffient (B1) is 0. After the hypothesis test is over, we see the p-value is very low and hence we reject the null hypothesis.

We learnt about p-values and coefficients. What is this t-statistic?
This is a measure of the number of standard deviations that estimate (coefficient B1 with a hat) away from zero.

t = (B1_hat - 0)/ Standard_error(B1_hat).

[Standard_error](https://github.com/chidamnat/practicalDS/blob/master/mastery/central_limit_theorem.ipynb) is the standard deviation of this test statistic.

More the t-statistic value, farther the coefficient from 0 and unlikely the null hypothesis be True, which is good for us so we can predict the sales using TV as a predictor

### Is this good enough ?
let us try adding another variable radio into the model.


In [0]:
result2 = smf.ols('sales ~ TV + radio', data=df).fit()
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Sun, 15 Dec 2019   Prob (F-statistic):           4.83e-98
Time:                        15:32:37   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9211      0.294      9.919      0.000       2.340       3.502
TV             0.0458      0.001     32.909      0.000       0.043       0.048
radio          0.1880      0.008     23.382      0.000       0.172       0.204
==============================================================================
Omnibus:                       60.022   Durbin-Watson:                   2.081
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              148.679
Skew:                          -1.323   Prob(JB):                     5.19e-33
Kurtosis:                       6.292   Cond. No.                         425.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

voila! good improvement to R-squared and now it is closer to 89.7%.

Notice that the TV's coefficient estimate is smaller now than when it was used as a sole predictor. More on this later

In [0]:
# Let us do all in 

results = smf.ols('sales ~ TV + radio + newspaper', data=df).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sun, 15 Dec 2019   Prob (F-statistic):           1.58e-96
Time:                        15:32:38   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Thats a lot of numbers. We shall understand the important ones.



```
          	coef	 std err	t	    P>|t|	[0.025	0.975]
Intercept	  2.9389	0.312	 9.422	0.000	2.324	  3.554
TV	        0.0458	0.001	32.809	0.000	0.043	  0.049
radio	      0.1885	0.009	21.893	0.000	0.172	  0.206
newspaper	 -0.0010	0.006	-0.177	0.860	-0.013	0.011
```

Here we see p-values pretty low for TV and radio. Whereas the newspaper has high p-value so newspaper is clearly insignificant or this case, we cannot reject the null hypothesis! (as statisticans say).

OK. Is our model good now ?

Yes. but what is the point in adding newspaper to the model when it is not adding any improvements. Sure. We can settle down with TV and radio as good predictors of sales. If I were to run any marketing campaign, I would bet on these two channels. OK, wait! How sure are you about this model if this is correct ?

Well, we have explored the t-statistic and got the individual predictor p-values. Isn't that enough ?

Well may be just not enough yet :)
Why ?
what does individual coeffient say ?
for example, the coefficient TV as a predictor states that out of 100 times, 95 times we will observe the true coeffient of TV (which B1 without a hat!) to be in the Confidence range of [0.043	  0.049] of B1_hat. So why is that an issue now ?
Well there is a counter point that 5% of the time the true value will take these extreme values. So what ?

Adding more variables / predictors would only compound this effect 5% * 5% * ... so on. So when more predictors are involved, individual p-value does not  give us the absolute way to evaluate our model as whole,

[Savior F-statistic](https://en.wikipedia.org/wiki/Ronald_Fisher) comes to our rescue. 
F = ((TSS / RSS)/p)/(RSS/(n-p-1)))
TSS => Total sum of squres or total variance
RSS => unexplained sum of squares (also 1 - R-squared)
n => number of observations
p => number of predictors used

In our case the F-statistic is 570 which is quite larger than 1. Hence we can safely conclude the null Hypothesis (B1_hat = B2_hat = ... = Bp_hat = 0) does not hold true and can be rejected.

Note: when p > n, linear regression does not work with least squares method and even F-statistic wouldn't matter. In this case, we deal with this later by other means such as dimensionality reduction (more on that later).


## Why to use linear model and when not to use it ?
Linear model comes with linearity assumption that the true distribution is linear in nature with respect to the predictor variables.

In real world mostly the linearity assumption may fail. Then why to use the linear model ? 
It is easy to explain and also not a bad model go start with unless linearity assumptions are broken. 

In Machine learning, there is [no free lunch theorem](https://en.wikipedia.org/wiki/No_free_lunch_theorem), which states that there is no one silver bullet model which could solve all the problems. So the problem space is quite high and to simplify this we need to make some assumptions about the world. If these assumptions are rigid, then our model may not resonate with reality (commonly called as bias). If we make the models super flexible, then it may result in complexity and overfitting (if you ask the question that model has seen, it will answer that correctly but outside this, it will have tough time). The overfitting results in high variance. So one should tradeoff between the bias and variance to build a good enough model to resonate with real system. 

Back to problems with linear models:
1) Linearity assumption becomes false - what if the underlying data is not truly linear and in this case, linear model will give us a straight line and there will be high residual errors. To counter that we have options to include non-linear predictiors such as TV**2 or sqrt(TV) depending on what works best for us or to try other models such as Tree based or other non linear approaches

2) error terms are correlated - Generally this is the case of timeseries data which is auto-correlated. This means that yesterday's value of a stock will determine today's stock value for example. Sometimes we see such patterns outside the timeseries data such as in survey (where same persons from a family tend to have similar food habits, which could cause some kind of unexexpected correlation). Generally [Durbin Watson](https://www.statsmodels.org/dev/generated/statsmodels.stats.stattools.durbin_watson.html) test shows if the autocorrelation in error terms is present. In our case, we have a value of 2 and hence no such issues.

3) pattern in error terms - The error plot should be uniformly distributed and there should not be underlying patterns, famously called as [heteroscedasticity](https://en.wikipedia.org/wiki/Heteroscedasticity) (A tongue twister!). This can be countered by applying transformations such as log or sqrt on response variable Y (sales in this example)

4) Outliers - They tend to pull the best fit line up or down. We could counter this using imputation techniques or drop these data points

5) High leverage points - Extreme predictor values (example say TV or radio X values here in this problem). Sometimes this gets tricky when individually the TV or radio may not be different from the rest but a combination may be an extreme value. Can be quantified using [leverage statistic](https://stats.stackexchange.com/questions/318854/leverage-statistic-equation-explanation)

6) Collinearity - If the predictors themselves are correlated to each other then the outcome cannot be easily explained by individual effect of the predictors accurately. For example, if we are interested in predicting the response variable called health_index based on height, weight and bmi, then we see the variables weight/height is highly correlated with bmi and hence makes it difficult to predict the health_index as a linear combination of these 3 predictors. To counter that we could use a correlation table to find these unwanted correlation among the predictor variables. But the problem becomes tricky if multiple variables are correlated (more than 2 variables). This is where VIF (Variance Inflation factor) comes to our rescue to detect the same. However for fixing it we need to drop one of the variables / reduce dimensionality to a much simpler representations (more on that later). 

We shall see how we can compute vif using statsmodels library as below

In [0]:
# vif example:
from patsy import dmatrices
from statsmodels.stats.outliers_influence import variance_inflation_factor
import statsmodels.api as sm

y, X = dmatrices("sales ~ TV + radio + newspaper ", data=df, return_type="dataframe")

# For each Xi, calculate VIF
vif = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

# Fit X to y
result = sm.OLS(y, X).fit()
print(vif)
## We are good in this case. Not much of multi-collinearity problem as the vaues are all closer to 1. 
# (high values of > 100 indicates multi-collinearity problems)

[6.848899953334954, 1.00461078493965, 1.1449519171055353, 1.1451873787239288]
